File: src/png.rs
//! Minimal grayscale PNG encoder (bit depth 8, color type 0), mirroring
//! pxpipe's png.ts: IHDR + one IDAT (zlib) + IEND, filter byte 0 per row.

use miniz_oxide::deflate::compress_to_vec_zlib;

const fn crc_table() -> [u32; 256] {
    let mut table = [0u32; 256];
    let mut n = 0;
    while n < 256 {
        let mut c = n as u32;
        let mut k = 0;
        while k < 8 {
            c = if c & 1 != 0 {
                0xedb8_8320 ^ (c >> 1)
            } else {
                c >> 1
            };
            k += 1;
        }
        table[n] = c;
        n += 1;
    }
    table
}

static TABLE: [u32; 256] = crc_table();

fn chunk(out: &mut Vec<u8>, ty: &[u8; 4], data: &[u8]) {
    out.extend_from_slice(&(data.len() as u32).to_be_bytes());
    out.extend_from_slice(ty);
    out.extend_from_slice(data);
    let mut c = 0xffff_ffffu32;
    for &b in ty.iter().chain(data.iter()) {
        c = TABLE[((c ^ b as u32) & 0xff) as usize] ^ (c >> 8);
    }
    out.extend_from_slice(&(!c).to_be_bytes());
}

pub fn encode_gray_png(fb: &[u8], w: usize, h: usize) -> Vec<u8> {
    debug_assert_eq!(fb.len(), w * h);
    let mut raw = Vec::with_capacity(h * (w + 1));
    for y in 0..h {
        raw.push(0); // filter: none
        raw.extend_from_slice(&fb[y * w..(y + 1) * w]);
    }
    let compressed = compress_to_vec_zlib(&raw, 6);
    let mut out = Vec::with_capacity(compressed.len() + 64);
    out.extend_from_slice(&[0x89, b'P', b'N', b'G', 0x0d, 0x0a, 0x1a, 0x0a]);
    let mut ihdr = [0u8; 13];
    ihdr[0..4].copy_from_slice(&(w as u32).to_be_bytes());
    ihdr[4..8].copy_from_slice(&(h as u32).to_be_bytes());
    ihdr[8] = 8; // bit depth
    ihdr[9] = 0; // grayscale
    chunk(&mut out, b"IHDR", &ihdr);
    chunk(&mut out, b"IDAT", &compressed);
    chunk(&mut out, b"IEND", &[]);
    out
}

File: src/png.ts
//! Minimal grayscale PNG encoder (bit depth 8, color type 0), mirroring
//! pxpipe's png.ts: IHDR + one IDAT (zlib) + IEND, filter byte 0 per row.

import { deflateSync } from "node:zlib";

const TABLE: Uint32Array = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = (c & 1) !== 0 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c >>> 0;
  }
  return table;
})();

/// Append a chunk (len BE + type + data + CRC over type||data) at `pos`.
function chunk(out: Uint8Array, pos: number, ty: string, data: Uint8Array): number {
  const len = data.length;
  out[pos] = (len >>> 24) & 0xff;
  out[pos + 1] = (len >>> 16) & 0xff;
  out[pos + 2] = (len >>> 8) & 0xff;
  out[pos + 3] = len & 0xff;
  const tyOff = pos + 4;
  for (let i = 0; i < 4; i++) out[tyOff + i] = ty.charCodeAt(i);
  out.set(data, tyOff + 4);
  let c = 0xffffffff;
  for (let i = tyOff; i < tyOff + 4 + len; i++) {
    c = TABLE[(c ^ out[i]) & 0xff] ^ (c >>> 8);
  }
  c = ~c >>> 0;
  const crcOff = tyOff + 4 + len;
  out[crcOff] = (c >>> 24) & 0xff;
  out[crcOff + 1] = (c >>> 16) & 0xff;
  out[crcOff + 2] = (c >>> 8) & 0xff;
  out[crcOff + 3] = c & 0xff;
  return crcOff + 4;
}

/// Encode a `w x h` grayscale framebuffer as an 8-bit color-type-0 PNG.
export function encodeGray(pixels: Uint8Array, w: number, h: number): Uint8Array {
  const raw = new Uint8Array(h * (w + 1));
  for (let y = 0, o = 0, s = 0; y < h; y++) {
    raw[o++] = 0; // filter: none
    raw.set(pixels.subarray(s, s + w), o);
    o += w;
    s += w;
  }
  const compressed: Uint8Array = deflateSync(raw, { level: 6 });
  const out = new Uint8Array(8 + 25 + (12 + compressed.length) + 12);
  out[0] = 0x89;
  out[1] = 0x50; // P
  out[2] = 0x4e; // N
  out[3] = 0x47; // G
  out[4] = 0x0d;
  out[5] = 0x0a;
  out[6] = 0x1a;
  out[7] = 0x0a;
  const ihdr = new Uint8Array(13);
  ihdr[0] = (w >>> 24) & 0xff;
  ihdr[1] = (w >>> 16) & 0xff;
  ihdr[2] = (w >>> 8) & 0xff;
  ihdr[3] = w & 0xff;
  ihdr[4] = (h >>> 24) & 0xff;
  ihdr[5] = (h >>> 16) & 0xff;
  ihdr[6] = (h >>> 8) & 0xff;
  ihdr[7] = h & 0xff;
  ihdr[8] = 8; // bit depth
  ihdr[9] = 0; // grayscale
  let pos = chunk(out, 8, "IHDR", ihdr);
  pos = chunk(out, pos, "IDAT", compressed);
  chunk(out, pos, "IEND", new Uint8Array(0));
  return out;
}
